hax/treemakers/lones2info.py

```diff
@@ -0,0 +1,228 @@
+
+
+def get_small_events(data):
+    """Return rows from data corresponding to events
+    that have S2 < 500 and S1 < 70 (or lack S1s altogether).
+    """
+    d = data[data['s2_area'] < 500]
+
+    # Remove a very large population of events with large S1s
+    # where the S2 is probably just fake nonsense after the S1 (AP, PI SE)
+    d = d[
+        np.isnan(d['s1_area'])
+        | (d['s1_area'] < 70)]
+
+    return d
+
+
+class BigBeforeSmall(hax.minitrees.MultipleRowExtractor):
+    """Return very basic info about all large S2s closely preceding a small-S2 event
+    "Large" means: S2 area > max(300 PE, 0.1 * largest_s2_area)
+    "Small" is defined by get_small_events
+    """
+    __version__ = '0.0.7'
+    extra_branches = ['peaks.reconstructed_positions*',
+                      'peaks.left', 'peaks.center_time']
+
+    def get_data(self, dataset, event_list=None):
+        if event_list is not None:
+            raise RuntimeError("BigBeforeSmall computes its own event list")
+
+        # Notice we use bypass_blinding here: we want *all* large events
+        # before small ones, even those that are candidates for funny processes
+        # like DEC and DBD.
+        # (looking back at this, this is probably not necessary, if I recall correctly;
+        # blinding only gets activated if you load the correction minitree
+        # which we do not)
+        data = hax.minitrees.load_single_dataset(
+            dataset,
+            ['Fundamentals', 'LargestPeakProperties'],
+            bypass_blinding=True)[0]
+
+        small_events = get_small_events(data)
+
+        # Find all events for which a small event follows in 1s
+        # This is about 10% of all events in a normal background dataset
+        dts = (small_events['event_time'].values.reshape(1, -1) -
+               data['event_time'].values.reshape(-1, 1))
+        close_mask = np.any((0 < dts) & (dts < int(1e9)), axis=1)
+        event_list = np.where(close_mask)[0]
+
+        return hax.minitrees.MultipleRowExtractor.get_data(self, dataset, event_list=event_list)
+
+    def extract_data(self, event):
+        # Get the largest S2s (up to five) satisfying
+        #   area > max(300 PE, 0.1 * largest_s2_area)
+        s2s = [p
+               for p in event.peaks
+               if p.type == 's2']
+        if not len(s2s):
+            # Hm, that's somewhat interesting...
+            # Maybe have a look at these sometime.
+            return []
+
+        s2_areas = np.array([p.area for p in s2s])
+        order = np.argsort(s2_areas)[::-1]
+        s2s = [s2s[j] for j in order]
+
+        largest_s2_area = s2_areas.max()
+        s2s = [p
+               for p in s2s
+               if p.area > max(300, largest_s2_area * 0.1)][:5]
+
+        result = []
+        for i, p in enumerate(s2s):
+            for rp in p.reconstructed_positions:
+                if rp.algorithm == 'PosRecTopPatternFit':
+                    x, y = rp.x, rp.y
```

Comment:
I think it's necessary to add the neural net pos which is the "official choice" in SR1 paper. And probably tensorflow position is helpful for some purpose.

Reply:
Good idea: I've added the neural net position. We've been using TPF so far, but we should be able to compare them. 

From https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenon1t:analysis:sr1:posrec#why_nn, it sounds like NN was favoured for better performance near dead PMTs and smoother overall response. In contrast TPF should work better on very low S2-events, which is the most important for us.

Reply:
I think we are (at least I'm) missing a plot to support the argument that TPF works better in the lowest energy, though this has nothing to do with this PR.